controller/kubernetes_configmap_controller.go

```diff
@@ -0,0 +1,248 @@
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	kc := &KubernetesConfigMapController{
+		baseController: newBaseController("longhorn-kubernetes-configmap-controller", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-kubernetes-configmap-controller"}),
+
+		cfmStoreSynced: configMapInformer.Informer().HasSynced,
+	}
+
+	configMapInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    kc.enqueueConfigMapChange,
+		UpdateFunc: func(old, cur interface{}) { kc.enqueueConfigMapChange(cur) },
+		DeleteFunc: kc.enqueueConfigMapChange,
+	})
+
+	return kc
+}
+
+func (kc *KubernetesConfigMapController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	kc.logger.Infof("Start")
+	defer kc.logger.Infof("Shutting down")
+
+	if !controller.WaitForCacheSync(kc.name, stopCh, kc.cfmStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesConfigMapController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesConfigMapController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesConfigMapController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		kc.logger.WithError(err).Warnf("Error syncing ConfigMap %v", key)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	kc.logger.WithError(err).Warnf("Dropping ConfigMap %v out of the queue", key)
+	kc.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (kc *KubernetesConfigMapController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", kc.name, key)
+	}()
+
+	namespace, cfmName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if err := kc.reconcileDefaultStorageClass(namespace, cfmName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (kc *KubernetesConfigMapController) reconcileDefaultStorageClass(namespace, cfmName string) error {
+	if namespace != kc.namespace || cfmName != types.DefaultStorageClassConfigMapName {
+		return nil
+	}
+
+	storageCFM, err := kc.ds.GetConfigMap(kc.namespace, types.DefaultStorageClassConfigMapName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			// The StorageClass ConfigMap was deleted therefore we delete the actual StorageClass
+			err := kc.ds.DeleteStorageClass(types.DefaultStorageClassName)
```

Comment:
I might be missing something, but I feel like this is a bad idea, since we won't be able to provision any volumes afterwards.

Reply:
Yes, you are right! Use will not be able to provision new volume with the `longhorn` StorageClass if we delete the `longhorn` storageclass.

This is the result when users manually delete the longhorn storage ConfigMap.  I feel that it is expected because the idea is that Longhorn will sync  longhorn storage ConfigMap with the actual `longhorn` StorageClass.

How do you think?

Reply:
It's a good thought, though
I feel like in the case where a config map (conceptually these are user defined settings for me) is not present we should:
- check for the existence of a longhorn storage class
-- if a class exists, do nothing
-- if no class exists, create a default class with default settings (we already have them defined in the longhorn ui)

Absence of dynamic user configuration should not lead to total system failure.
What do you think?

Reply:
For me the config map is conceptually an override
layer0: default settings
layer1: user settings (config-map)
absence of layer1 should not lead to failure

Reply:
Yeah, good point!

However, I would prefer to completely separate the storageclass setting from UI default setting. Right now, some users are confused about those two since what they change in UI doesn't reflect on the actually `longhorn` storageclass. We have been telling them that changing UI setting has nothing to do with the volumes created via CSI using `longhorn` storageclass

Moreover, if users want to delete `longhorn` storageclass? If we fall back to default settings and recreate a new one, it means that users never be able to remove `longhorn` storageclass unless they remove the setting CRDs.

Reply:
@yasker, @PhanLe1010 and me had a good design conversation and we explored only doing this during upgrade but in the end decided to do it at runtime and treat the config map as the source of truth for the default longhorn storage class.

But we agreed no to do anything if the config map is absent.

Reply:
Thanks @joshimoo !

Update the PR to do nothing (not punish users :) ) if the config map is absent.